types/next-auth/client.d.ts

```diff
@@ -46,6 +53,14 @@ declare function session(
 declare const getSession: typeof session;
 declare function csrfToken(context?: NextContext): Promise<string | null>;
 declare const getCsrfToken: typeof csrfToken;
+declare function signin(
+    provider: 'credentials' | 'email' | 'string',
```

Comment:
Lol I think it's my issue 😂 as I suggested it [wrongly here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51614#discussion_r589066083)

Reply:
I think just `'credentials' | 'email'`.

I already changed

Reply:
the provider can really be anything. either one of the built-in ones, email or credentials. only  email and credentials won't be sufficient. example from docs https://next-auth.js.org/getting-started/client#starts-google-oauth-sign-in-flow-when-clicked

Reply:
Hey @balazsorban44

Yes! I got it. 
But what I'm doing is an overlap. In this case, only will return `SignInResponse` when the provider is `credentials` or `email`. Look these lines [here](https://github.com/nextauthjs/next-auth/blob/main/src/client/index.js#L212) and [here](https://github.com/nextauthjs/next-auth/blob/main/src/client/index.js#L235)

If the provider will be anything else the TS will use another declare, and this another declare will check the type `string` [as you see here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51614/files#diff-16239debf32a8cdbd7559e8e5cdb55d204f46cc804c0a1523025fcd5d0f27144R65)

Reply:
@guilouro I downloaded this change and looks like `provider` should be `string` instead of `'credentials' | 'email'`. Because I use custom provider and the first argument I pass is `'credentials-register'` for example. First param matches provider ID which can be any string.

Regarding your comment that TS will use another declare: why if type of `redirect` is different?

cc: @balazsorban44

Reply:
Hey @yuzhakovvv, sorry for my delay to answer. 

Then, TS will use declare with `Promise<SignInResponse>` return only if the provider is `'credentials' | 'email'` with `redirect` false. With any value else TS will use declare with `Promise<void>`

Reply:
@guilouro isn't it expected?

As an alternative solution maybe we leave a single declare with return `Promise<SignInResponse | null>`?

I'm just looking for a solution for my case with custom provider ID. Restriction to `'credentials' | 'email'` doesn't allow custom provider IDs even though they use Email or Credentials providers. Do you have another idea of how to solve this?

Reply:
@guilouro @balazsorban44 any thoughts? 🤓